src/controllers/GovernanceTimelock.sol

```diff
@@ -0,0 +1,142 @@
+                            STATE VARIABLES
+    ////////////////////////////////////////////////////////////////*/
+    uint256 public txCounter;
+    uint256 public delay = MINIMUM_DELAY;
+
+    /// @dev the admin should be multi-message receiver
+    address public multiMessageReceiver;
+
+    mapping(uint256 txId => ScheduledTransaction) public scheduledTransaction;
+    mapping(uint256 txId => bool executed) public isExecuted;
+
+    /*/////////////////////////////////////////////////////////////////
+                                MODIFIERS
+    ////////////////////////////////////////////////////////////////*/
+
+    /// @notice A modifier used for restricting the caller of some functions to be this contract itself.
+    modifier onlySelf() {
+        if (msg.sender != address(this)) {
+            revert Error.INVALID_SELF_CALLER();
+        }
+        _;
+    }
+
+    /// @notice A modifier used for restricting caller to mma receiver contract
+    modifier onlyMultiMessageReceiver() {
+        if (msg.sender != multiMessageReceiver) {
+            revert Error.CALLER_NOT_MULTI_MESSAGE_RECEIVER();
+        }
+        _;
+    }
+
+    /*/////////////////////////////////////////////////////////////////
+                                CONSTRUCTOR
+    ////////////////////////////////////////////////////////////////*/
+
+    /// @param _multiMessageReceiver is the address of multiMessageReceiver
+    constructor(address _multiMessageReceiver) {
+        if (_multiMessageReceiver == address(0)) {
+            revert Error.ZERO_ADDRESS_INPUT();
+        }
+
+        multiMessageReceiver = _multiMessageReceiver;
+    }
+
+    /*/////////////////////////////////////////////////////////////////
+                            EXTERNAL FUNCTIONS
+    ////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IGovernanceTimelock
+    function scheduleTransaction(address _target, uint256 _value, bytes memory _data)
+        external
+        override
+        onlyMultiMessageReceiver
+    {
+        if (_target == address(0)) {
+            revert Error.INVALID_TARGET();
+        }
+
+        /// increment tx counter
+        ++txCounter;
+        uint256 eta = block.timestamp + delay;
+
+        scheduledTransaction[txCounter] = ScheduledTransaction(_target, _value, _data, eta);
+
+        emit TransactionScheduled(txCounter, _target, _value, _data, eta);
+    }
+
+    /// @inheritdoc IGovernanceTimelock
+    function executeTransaction(uint256 txId) external override {
+        /// @dev validates the txId
+        if (txId == 0 || txId > txCounter) {
+            revert Error.INVALID_TX_ID();
+        }
+
+        /// @dev checks if tx is already executed;
+        if (isExecuted[txId]) {
+            revert Error.TX_ALREADY_EXECUTED();
+        }
+
+        ScheduledTransaction memory transaction = scheduledTransaction[txId];
+
+        /// @dev checks timelock
+        if (transaction.eta < block.timestamp) {
```

Comment:
we can simplify this implementation by having a GRACE_PERIOD constant, that we check against inside `executeTransaction()`. i.e. we ensure that `eta + GRACE_PERIOD > block.timestamp` (see updated spec notion page for details)

Reply:
Each message already has an `expiration` field. So `GRACE_PERIOD` may not be useful if the `expiration` field is properly set, though I agree that `GRACE_PERIOD` may be nice to have for the completeness of the timelock contract itself.

Reply:
never mind, I got confused. `expiration` is only meaningful before the message is scheduled. So my comment above is incorrect.

Reply:
The terms now may be a bit confusing, as we have msg execution (= tx scheduling) and tx execution. The `ExecutionData.expiration` value initiated by the message sender only specifies the deadline for the tx to be scheduled.

Reply:
Yeah, there are two aspects to this: 
1. Should the timelock enforce some check on staleness or should we just rely on the expiration field being enforced by the receiver?
    Relying solely on the message expiration field enforced by the MultiMessageReceiver could result in the edge case we previously discussed. Consider the following scenario:
    - Transaction tx1 passes the timelock, but its execution fails because of some condition X.
    - The transaction is then abandoned in favour of a subsequent rectifying transaction, tx2.
    - However, if at a future point in time, the condition (X) that caused tx1 to fail in the past no longer holds, then anyone could execute tx1, potentially causing undesirable behaviour.

2. If the timelock should enforce a staleness check, should it use the message expiration field or its own grace period?
Although it may seem redundant, there are a couple of advantages to having a separate grace period in the timelock, as you mentioned. It facilitates rolling out to deployments that do not currently use the message expiration validation logic (and will not use MMA, such as L2 deployments), reduces the potential for errors by proposers who would have to account for chain-specific timelock delays when setting message expiration values and provides flexibility in separating delivery time bounds from execution time bounds.